misc/db-updates/2023-07-17/creatorsAsNonClassicDataObjects

```diff
@@ -0,0 +1,71 @@
+
+Z_Core::$debug = true;
+
+$shardIDs = Zotero_DB::columnQuery("SELECT shardID FROM shards WHERE shardHostID=? AND shardID >= ? AND shardID <= ? ORDER BY shardID", [$shardHostID, $startShard, $stopShard]);
+foreach ($shardIDs as $shardID) {
+	echo "Shard: $shardID\n";
+
+	echo "Setting shard to readonly\n";
+	Zotero_DB::query("UPDATE shards SET state='readonly' WHERE shardID=?", $shardID);
+	
+	echo "Waiting 60 seconds for requests to stop\n";
+	sleep(60);
+	
+	// Creators 
+	echo "Migrating creators\n";
+	// Drop foreign key constraint
+	Zotero_Admin_DB::query("ALTER TABLE `itemCreators` DROP CONSTRAINT `itemCreators_ibfk_1`;", false, $shardID);
+	Zotero_Admin_DB::query("ALTER TABLE `itemCreators` DROP CONSTRAINT `itemCreators_ibfk_2`;", false, $shardID);
+
+	// Create new itemCreators table
+	Zotero_Admin_DB::query("CREATE TABLE `itemCreatorsNew` ( `creatorID` BIGINT UNSIGNED NOT NULL, `itemID` BIGINT UNSIGNED NOT NULL, `firstName` varchar(255) CHARACTER SET utf8mb4 COLLATE utf8mb4_unicode_ci DEFAULT NULL, `lastName` varchar(255) CHARACTER SET utf8mb4 COLLATE utf8mb4_unicode_ci DEFAULT NULL, `fieldMode` tinyint(1) UNSIGNED DEFAULT NULL, `creatorTypeID` smallint(5) UNSIGNED NOT NULL, `orderIndex` smallint(5) UNSIGNED NOT NULL, PRIMARY KEY (`creatorID`, `itemID`), KEY `creatorTypeID` (`creatorTypeID`), KEY `name` (`lastName`(7),`firstName`(6)) ) ENGINE=InnoDB DEFAULT CHARSET=utf8;", false, $shardID);
+
+	// Add foreign key to item constraint
+	Zotero_Admin_DB::query("ALTER TABLE `itemCreatorsNew` ADD CONSTRAINT `itemCreators_ibfk_1` FOREIGN KEY (`itemID`) REFERENCES `items` (`itemID`) ON DELETE CASCADE;", false, $shardID);
+
+	// Populate new table with data
+	Zotero_Admin_DB::query("INSERT INTO itemCreatorsNew (creatorID, firstName, lastName, fieldMode, itemID, creatorTypeID, orderIndex ) SELECT creatorID, firstName, lastName, fieldMode, itemID, creatorTypeID, orderIndex from creators INNER JOIN itemCreators USING (creatorID);", false, $shardID);
+
+	// Drop old creators tables
+	Zotero_Admin_DB::query("DROP TABLE itemCreators;", false, $shardID);
+	Zotero_Admin_DB::query("DROP TABLE creators;", false, $shardID);
+
+	// Rename old itemCreators table
+	Zotero_Admin_DB::query("RENAME TABLE itemCreatorsNew TO itemCreators;", false, $shardID);
+
+	// Tags
+	echo "Migrating tags\n";
+	// Drop foreign key constraint
+	Zotero_Admin_DB::query("ALTER TABLE `itemTags` DROP CONSTRAINT `itemTags_ibfk_1`;", false, $shardID);
+	Zotero_Admin_DB::query("ALTER TABLE `itemTags` DROP CONSTRAINT `itemTags_ibfk_2`;", false, $shardID);
+
+	// Create new itemTags table
+	Zotero_Admin_DB::query("CREATE TABLE `itemTagsNew` ( `tagID` BIGINT UNSIGNED NOT NULL, `itemID` BIGINT UNSIGNED NOT NULL, `name` varchar(255) CHARACTER SET utf8mb4 COLLATE utf8mb4_bin NOT NULL, `type` tinyint(1) unsigned NOT NULL DEFAULT '0', `version` int(10) unsigned NOT NULL DEFAULT '1', PRIMARY KEY (`tagID`, `itemID`) ) ENGINE=InnoDB DEFAULT CHARSET=utf8;", false, $shardID);
```

Comment:
1. Yes, there is a [test](https://github.com/zotero/dataserver/blob/eb3fd636ec0f50eb0bc2ecc310ca1db3e77ed9b1/tests/remote/tests/API/3/TagTest.php#L687) that calls `/tags?newer=`, so I assumed we need to keep it around. If we don't want it - that's great, I'll remove it from here and from my mocha tests PR.

Reply:
2. Yeah, some context from real world data would certainly be helpful. Maybe if we could see the output of `select count(*) from itemTags` and divide it by `select count(*) from tags`, we'd get an idea of roughly many `itemTag` relationships there are for each tag. And then if we multiply that by `select SUM(length(name)) from tags`, we'd have a rough estimate of how much extra memory duplicating tag names would cause.

Reply:
3. Queries related to tags that were affected are:

- In `/items?tag=` where we select the itemIDs that fit the tag filter `SELECT itemID FROM items JOIN itemTags USING (itemID) WHERE LOWER(itemTags.name) IN (?)`

- In `tags/?tag=` where we select tagIDs that fit the query before fetching their actual data and constructing objects  `SELECT SQL_CALC_FOUND_ROWS DISTINCT tagID FROM itemTags JOIN items USING (itemID) WHERE libraryID=? AND `name` in (?) AND `type` in (?)`;

Currently, these do a full table scan, which is not good. There was a `UNIQUE KEY uniqueTags (libraryID, name, type)` in the old creators table that was used for such queries. We can add `KEY nameType (name, type)` to current `itemTags` and these queries will use that key. I checked manually with `explain` sql statement and it seems to work.

With creators, the `name` key indeed does not do anything now because the creators never fetched by themselves and are only loaded as a part of an item via `SELECT * FROM itemCreators WHERE itemID=?`. 

To sum up, I guess we should add `KEY nameType (name, type)` to itemTags and remove `name` from creators?